pkg/network/ovn_kubernetes.go

```diff
@@ -2303,6 +2488,58 @@ func isInterConnectEnabledOnMasterDaemonset(ds *appsv1.DaemonSet) bool {
 	return isInterConnectEnabledOnDaemonset(ds, util.OVN_MASTER)
 }
 
+func isOVNIPsecNotActive(ds *appsv1.DaemonSet) bool {
+	// If no daemonset, then return false.
+	if ds == nil {
+		return false
+	}
+	// When observed generation doesn't match with spec generation or in progressing state
+	// then return false as we are not sure about IPsec state.
+	if ds.Generation != ds.Status.ObservedGeneration || daemonSetProgressing(ds, true) {
+		return false
+	}
+	annotations := ds.GetAnnotations()
+	// If OVN daemonset is set with IPsecEnableAnnotation, then return false.
+	if annotations[names.IPsecEnableAnnotation] != "" {
+		return false
+	}
+	// If IPsec is running with older version and ipsec=true is found from OVN daemonset, then return true.
+	if annotations["release.openshift.io/version"] != os.Getenv("RELEASE_VERSION") &&
+		isIPSecEnabledInDaemonSet(ds, util.OVN_NBDB) {
+		return false
+	}
+	// All other cases, return true.
+	return true
+}
+
+func isIPSecEnabledInDaemonSet(ds *appsv1.DaemonSet, containerName string) bool {
+	for _, container := range ds.Spec.Template.Spec.Containers {
+		if container.Name == containerName {
+			for _, c := range container.Lifecycle.PostStart.Exec.Command {
+				if strings.Contains(c, "ipsec=true") {
+					return true
+				}
+			}
+			break
+		}
+	}
+	return false
+}
+
+func isIPSecActiveInMasterStatefulSet(ss *appsv1.StatefulSet) bool {
+	for _, container := range ss.Spec.Template.Spec.Containers {
+		if container.Name == util.OVN_NBDB {
+			for _, c := range container.Lifecycle.PostStart.Exec.Command {
+				if strings.Contains(c, "ipsec=true") {
+					return true
+				}
+			}
+			break
+		}
+	}
+	return false
+}
+
```

Comment:
So looking at this, I think we should set the annotation in the pod spec within the daemon set as well.
Then you only need a single method instead of three that receives the pods spec `isOVNIPsecNotActive(pod *corev1.Pod)`. You then pass in to this method the pod spec from the daemon set or the stateful set depending on the version you are upgrading from.